File: 0049-group-anagrams/0049-group-anagrams.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        if (strs.empty()) return {};
        
        unordered_map<string, vector<string>> ansMap;
        vector<int> count(26, 0);
        
        for (string s : strs) {
            fill(count.begin(), count.end(), 0);
            
            for (char c : s) {
                count[c - 'a']++;
            }
            
            string key = "";
            for (int i = 0; i < 26; i++) {
                key += "#" + to_string(count[i]);
            }
            
            ansMap[key].push_back(s);
        }
        
        vector<vector<string>> result;
        for (auto& it : ansMap) {
            result.push_back(it.second);
        }
        return result;
    }
};

File: 0049-group-anagrams/0049-group-anagrams.java
import java.util.*;

class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        if (strs == null || strs.length == 0) 
            return new ArrayList<>();

        Map<String, List<String>> ansMap = new HashMap<>();
        int[] count = new int[26];

        for (String s : strs) {
            Arrays.fill(count, 0);

            for (char c : s.toCharArray()) {
                count[c - 'a']++;
            }

            StringBuilder keyBuilder = new StringBuilder();
            for (int i = 0; i < 26; i++) {
                keyBuilder.append('#');
                keyBuilder.append(count[i]);
            }

            String key = keyBuilder.toString();
            ansMap.computeIfAbsent(key, k -> new ArrayList<>()).add(s);
        }

        return new ArrayList<>(ansMap.values());
    }
}
